src/pygeoprocessing/kernels.py

```diff
@@ -0,0 +1,345 @@
+      tiled.
+    * The pixel values of a kernel may not represent nodata (GDAL's internal
+      way of indicating invalid pixel data)
+    * Pixel values must be finite, real numbers.  That is, they may not
+      represent positive or negative infinity or NaN.
+
+The functions in this module represent kernels that we have found to be
+commonly used, including:
+
+    * Dichotomous kernel: :meth:`dichotomous_kernel`
+    * Gaussian (normal) decay kernel: :meth:`normal_distribution_kernel`
+    * Linear decay kernel: :meth:`linear_decay_kernel`
+    * Exponential decay kernel: :meth:`exponential_decay_kernel`
+
+Additionally, the user may define their own kernel using the helper functions
+included here:
+
+    * :meth:`kernel_from_numpy_array`, for kernels where the kernel array is
+      already available as a ``numpy`` array, such as those used in common
+      image processing operations like sharpening and edge detection.
+      https://en.wikipedia.org/wiki/Kernel_(image_processing)
+    * :meth:`create_distance_decay_kernel`, for kernels that are a function of
+      distance from the center pixel
+"""
+import logging
+import math
+from typing import Callable
+from typing import Text
+from typing import Union
+
+import numpy
+import pygeoprocessing
+from numpy.typing import ArrayLike
+from osgeo import gdal
+
+FLOAT32_NODATA = float(numpy.finfo(numpy.float32).min)
+LOGGER = logging.getLogger(__name__)
+
+
+def kernel_from_numpy_array(
+        numpy_array: ArrayLike, target_kernel_path: Text) -> None:
+    """Create a convolution kernel from a numpy array.
+
+    Args:
+        numpy_array: A 2-dimensional numpy array to convert into a raster.
+        target_kernel_path: The path to where the kernel should be written.
+
+    Returns:
+        ``None``
+    """
+    n_dimensions = numpy.array(numpy_array).ndim
+    if n_dimensions != 2:
+        raise ValueError(
+            "The kernel array must have exactly 2 dimensions, not "
+            f"{n_dimensions}")
+
+    # The kernel is technically a TIFF, not a GeoTiff.
+    pygeoprocessing.numpy_array_to_raster(
+        numpy_array, target_nodata=None, pixel_size=None, origin=None,
+        projection_wkt=None, target_path=target_kernel_path)
+
+
+def dichotomous_kernel(
+        target_kernel_path: Text,
+        max_distance: Union[int, float],
+        apothem: Union[int, float] = None,
+        normalize: bool = True) -> None:
+    """Create a binary kernel indicating presence/absence within a distance.
+
+    This is equivalent to ``int(dist <= max_distance)`` for each pixel in the
+    kernel, where ``dist`` is the euclidean distance of the pixel's centerpoint
+    to the centerpoint of the center pixel.
+
+    Args:
+        target_kernel_path: The path to where the kernel will be written.
+            Must have a file extension of ``.tif``.
+        max_distance: The maximum distance of the kernel, in pixels. Kernel
```

Comment:
Sorry, this wording is still tripping me up... what about something like
```suggestion
        max_distance: The distance threshold, in pixels. Kernel
```